# Computational Theory - Jamie Walsh

## Contents

- [Problem 1: Representing SHA-256 data](#problem-1-representing-sha-256-data)
- [Problem 2: SHA-256 Bitwise Operations](#problem-2-sha-256-bitwise-operations)

In [4]:
import numpy as np
import sys
import struct

## Problem 1: Representing SHA-256 data

Before beginning any work on implementing SHA-256, this section will cover how we represent SHA-256's inputs, outputs and intermediate data in Python, and why these representations are appropriate.

### 32-bit words

Why 32 bits? SHA-256 requires 32-bit words, as specified in [FIPS 180-4 Section 3.2](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=13). Within that detail there's a constraint which is important for how we represent this data: values are limited to 32 bits and overflow if this limited is exceeded.

Unlike "lower-level" languages like C++, where a 32-bit integer comes as a built-in `uint32_t`, Python's `int` does not behave this way.

We need to use a data type that won't exceed 32 bits. To test this, we can take the max value and add 1. Python's plain `int` has no clean way to enforce this limit; however, [NumPy's `np.iinfo`](https://stackoverflow.com/questions/23189506/maximum-allowed-value-for-a-numpy-data-type) provides a solution.

In [5]:
limit = np.iinfo(np.uint32).max

print(f"int: {limit + 1}")  # plain int: no limit, keeps growing
print(f"np.uint32: {np.uint32(limit) + 1}")  # np.uint32: wraps back to 0

int: 4294967296
np.uint32: 0


/var/folders/3q/zzb1pzp9655bb6h57vz921yh0000gn/T/ipykernel_28606/1439755418.py:4: RuntimeWarning: overflow encountered in scalar add
  print(f"np.uint32: {np.uint32(limit) + 1}")  # np.uint32: wraps back to 0


This confirms we need `np.uint32` as it wraps the 32-bit limit while Python's plain `int` does not, which FIPS 180-4 requires.

### Sequences of 32-bit words
Also specified in [FIPS 180-4 Section 3.1](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=13), SHA-256 represents each 512-bit message block as a sequence of sixteen 32-bit words. To represent this, we use a NumPy array over a standard list, as [NumPy's documentation](https://numpy.org/doc/stable/user/whatisnumpy.html) states arrays are more efficient when every element is the same type, which applies here as every element is a `uint32`.

We can verify this claim in code:


In [6]:
numpy_array = np.array([1, 2, 3, 4, 5], dtype=np.uint32)
python_list = [np.uint32(x) for x in [1, 2, 3, 4, 5]]

print(f"Numpy Array Size: {numpy_array.nbytes}")  # array's total memory

# lists total memory: size of pointers + size of every object they point to
list_elements_size = sum(sys.getsizeof(x) for x in python_list)
list_size = sys.getsizeof(python_list) + list_elements_size
print(f"Python List Size: {list_size}")

print("numpy array:")
%timeit numpy_array + 1

print("python list:")
%timeit [x + 1 for x in python_list]

Numpy Array Size: 20
Python List Size: 260
numpy array:
477 ns ± 10.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
python list:
231 ns ± 0.579 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


From this data, the array (20 bytes) is much more efficient than the list (260 bytes) which supports NumPy's claims. However, we can see that the array was actually slower here (472ns vs 238ns) so why?

NumPy has a small, fixed overhead every time it performs an operation. With only 5 elements, that overhead outweighs any benefit. NumPy's speed advantage only shows up once arrays are large enough to outweigh this fixed cost.

So for our case, is this overhead actually an issue? Later problems use larger sequences, for example the 64-word message schedule in Problem 5. So we retest below at the same size, rather than assume this result generalises.


In [7]:
numpy_array = np.array(range(0, 64), dtype=np.uint32)
python_list = [np.uint32(x) for x in range(0, 64)]

print("numpy array:")
%timeit numpy_array + 1

print("python list:")
%timeit [x + 1 for x in python_list]

numpy array:
474 ns ± 2.17 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
python list:
2.46 μs ± 72 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


As predicted, retesting at a larger size (64 elements) shows the array is now significantly faster (486ns vs 2.44μs), roughly 5x faster. This confirms the array remains the right choice.

### Input Messages

Input messages can be stored as strings and converted to bytes using `.encode()` when required.

In [8]:
message = "abc"
message_bytes = message.encode()
print(message_bytes)

b'abc'


This confirms `.encode()` produces `bytes`, ready for further processing.

### 512-bit Message Blocks

A 512-bit message block is sixteen words. A block starts out as 64 bytes, so representing it as 16 words requires expressing those bytes as words. At a low level, these are all just bits, so rather than converting them (which costs computation), we can simply change how they're represented.

In [9]:
example_list = list(range(1, 17))

packed_bytes = struct.pack('>16I', *example_list)
# struct.unpack: reads raw bytes back as a tuple of ints, no computation
unpacked = struct.unpack('>16I', packed_bytes)
print(f"Round-trip correct: {unpacked == tuple(example_list)}")

elements_size = sum(sys.getsizeof(x) for x in example_list)
list_size = sys.getsizeof(example_list) + elements_size
print(f"List size: {list_size}")
print(f"Packed bytes size: {sys.getsizeof(packed_bytes)}")

example_bytes = b'\x00\x00\x00\x05'

print("manual bit-shifting:")
%timeit (example_bytes[0] << 24) | (example_bytes[1] << 16) | (example_bytes[2] << 8) | example_bytes[3]

print("struct.unpack:")
%timeit struct.unpack('>I', example_bytes)[0]

words_array = np.frombuffer(packed_bytes, dtype='>u4')
print(f"Words array: {words_array}")

Round-trip correct: True
List size: 632
Packed bytes size: 97
manual bit-shifting:
71.4 ns ± 0.902 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
struct.unpack:
48 ns ± 0.636 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
Words array: [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16]


The results confirm two things. First, the round trip check shows the conversion is lossless. Unpacking the packed bytes returns exactly the original sixteen numbers, so no information is lost by reinterpreting rather than computing. Second, the size and speed results show that reinterpreting bytes as words (type punning) is both more compact and faster than either storing the values in a list or manually reconstructing each word with bit-shifting

When reading through the problems, I remembered seeing clever time and cost saving measures with bits and bytes from the [Fast Inverse Square Root algorithm](https://down2core.com/docs/lomont.pdf). It uses the same underlying trick: reading the same bits through a different type, with no computation involved. The difference is what it's used for. Fast Inverse Square Root trades accuracy for speed, using this trick to get an approximate answer. Here, the same trick is used the opposite way, for an exact, lossless reformatting of data, since SHA-256 needs every value to be bit-exact

Also of note: Two format strings are used here `>16I` and `>u4`. Both start with `>` meaning big-endian, the reason this is important will be touched on more below.

### The final 256-bit hash value

[FIPS 180-4 Section 2.2.1](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=9) states that `H_0^(i)` "is the left-most word of hash value i". Since the hash value is 256 bits and a word is 32 bits, the hash value is exactly 8 words, numbered 0 to 7, left to right:

[0] [1] [2] [3] [4] [5] [6] [7]

In [10]:
hash_words = np.array(range(1, 9), dtype='>u4')  # 8 words matching [0] ... [7]
# words -> bytes, big-endian keeps word order matching [0] ... [7]
hash_bytes = hash_words.tobytes()
# bytes -> 64-character hex string, the final output format
hex_string = hash_bytes.hex()
print(hex_string)
print(len(hex_string))

0000000100000002000000030000000400000005000000060000000700000008
64


### Big-endian integers 

Back to the standard: [FIPS 180-4 Section 3.1](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=12) states that "Throughout this specification, the “big-endian” convention is used when expressing both 32- and 64-bit words, so that within each word, the most significant bit is stored in the left-most bit position".

So far we've been using `np.uint32` directly to represent a 32-bit word for arithmetic. Words, sequences, blocks and the hash value have all used it the same way without issue. This works everywhere except when converting to or from raw bytes. 

If we have a number `1234` and split it into `1`,`2`,`3`,`4`. We now have to decide which order to put them in, most significant first: (`1`,`2`,`3`,`4`) or least significant first: (`4`,`3`,`2`,`1`). The first is what "big-endian" means. The same applies to a `uint32`, converting it into 4 separate bytes splits it into pieces, and "which byte comes first" has to be consistent. A machine's native byte order may not answer the way FIPS 180-4 requires.

In [11]:
u32 = np.uint32  # native byte order -- safe for arithmetic
u32_be = '>u4'  # big-endian -- required when converting to/from raw bytes

Going forward we will use two short aliases to make this distinction more explicit: `u32` for ordinary arithmetic, and `u32_be` specifically for the moments a word needs to become, or come from, raw bytes, where big-endian must be stated rather than assumed.

### Summary of representation choices
- A **word** is `u32` (i.e. `np.uint32`), which enforces SHA-256's wraparound behavior
- A **sequence of words**, a **block** (16 words), and the **hash value** (8 words) are the same thing: a NumPy array of `u32` (i.e. `np.uint32`), with different lengths
- An **input message** is `bytes`, converted from a string with `.encode()`
- Bytes will be converted into words using **type punning** (`struct/np.frombuffer`) rather than manual computation
- **Big-endian** exists more as a rule for how conversions must be done rather than a type itself. `u32/u32_be` make this distinction explicit. 

## Problem 2: SHA-256 Bitwise Operations

SHA-256 will require some Boolean and bitwise functions for its actual implementation. These are seen in [FIPS 180-4 Section 4.1.2](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=15) as follows:
1. `Ch(x, y, z)`
2. `Maj(x, y, z)`
3. `Sigma0(x)` — written as Σ0{256}(x) in the standard
4. `Sigma1(x)` — written as Σ1{256}(x) in the standard
5. `sigma0(x)` — written as σ0{256}(x) in the standard
6. `sigma1(x)` — written as σ1{256}(x) in the standard

This section will cover each one's implementation, along with a brief explanation and test for each.

Since several of these functions take the same 3 boolean inputs (`x, y, z`), a shared testing setup is defined once here and reused for each, rather than repeating it per function. `run_truth_table` is generic and works for any number of inputs; `xyz` specifically enumerates all 8 combinations for a 3-input boolean function, reused by both `Ch` and `Maj` since they share that signature. Only the expected outputs differ per function.

In [12]:
xyz = [
    (0, 0, 0), (0, 0, 1), (0, 1, 0), (0, 1, 1),
    (1, 0, 0), (1, 0, 1), (1, 1, 0), (1, 1, 1),
]

def run_truth_table(func, inputs, expected_values):
    for values, expected in zip(inputs, expected_values):
        result = func(*values)
        print(
            f"{func.__name__}{values} = {result}, "
            f"expected {expected}: {result == expected}"
        )

### Ch(x, y, z)

The Ch function: $Ch(x,y,z) = (x \land y) \oplus (\lnot x \land z)$

Broken down in pseudocode this becomes:
```pseudo
for each bit i:
    if x[i] == 1:
        result[i] = y[i]
    else:
        result[i] = z[i]
```

In [13]:
def Ch(x: u32, y: u32, z: u32) -> u32:
    """Ch function: For each bit positon, select y's bit if x is 1,
    else z's bit."""
    return (x & y) ^ (~x & z)

In [14]:
ch_expected = [0, 1, 0, 1, 0, 0, 1, 1]
run_truth_table(Ch, xyz, ch_expected)

Ch(0, 0, 0) = 0, expected 0: True
Ch(0, 0, 1) = 1, expected 1: True
Ch(0, 1, 0) = 0, expected 0: True
Ch(0, 1, 1) = 1, expected 1: True
Ch(1, 0, 0) = 0, expected 0: True
Ch(1, 0, 1) = 0, expected 0: True
Ch(1, 1, 0) = 1, expected 1: True
Ch(1, 1, 1) = 1, expected 1: True


All 8 truth table combinations match, confirming `Ch` correctly implements the choose function: selecting `y`'s bit where `x = 1`, and `z`'s bit where `x = 0`.

### Maj(x, y, z)

The Maj function: $Maj(x,y,z) =  (x \land y) \oplus (x \land z) \oplus ( y \land z)$

Broken down in pseudocode this becomes:
```pseudo
for each bit i:
    count = x[i] + y[i] + z[i]
    if count is >= to 2:
        result[i] = 1
    else:
        result[i] = 0
```

In [15]:
def Maj(x: u32, y: u32, z: u32) -> u32:
    """Majority vote per bit: 1 if at least 2 of x, y, z are 1, else 0."""
    return (x & y) ^ (x & z) ^ (y & z)

In [16]:
maj_expected = [0, 0, 0, 1, 0, 1, 1, 1]
run_truth_table(Maj, xyz, maj_expected)

Maj(0, 0, 0) = 0, expected 0: True
Maj(0, 0, 1) = 0, expected 0: True
Maj(0, 1, 0) = 0, expected 0: True
Maj(0, 1, 1) = 1, expected 1: True
Maj(1, 0, 0) = 0, expected 0: True
Maj(1, 0, 1) = 1, expected 1: True
Maj(1, 1, 0) = 1, expected 1: True
Maj(1, 1, 1) = 1, expected 1: True


All 8 truth table combinations match, confirming `Maj` correctly implements the majority vote: the result is 1 whenever at least 2 of `x, y, z` are 1.

### ROTR(x, n)

[FIPS 180-4 Section 3.2](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=13) defines rotate right as `ROTR^n(x) = (x >> n) | (x << (32 - n))`: a normal right-shift combined with the same value shifted left, so bits that fall off the right appear at the top instead of being lost.

The implementation below is adapted from [this Stack Overflow thread](https://stackoverflow.com/questions/55856773/python-sha256-hash-computation), which shows a SHA-256 implementation producing incorrect output because its `ROTR` used plain Python `int`. Using `np.uint32` instead fixes this, since it enforces the 32-bit boundary automatically.

In [17]:
def ROTR(x: u32, n: int) -> u32:
    """Rotate x right by n bits."""
    return (x >> n) | (x << (32 - n))

### SHR(x, n)

[FIPS 180-4 Section 3.2](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=13) defines shift right as `SHR^n(x) = x >> n`: bits shift right by `n` positions, any bits that fall off are discarded. Python has a shift right operator which we already use above (`>>`), which makes this function a very simple one-liner.

In [18]:
def SHR(x: u32, n: u32) -> u32:
    """Shift x right by n bits."""
    return x >> n

### Sigma0(x) and Sigma1(x)

[FIPS 180-4 Section 4.1.2](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=15) defines `Sigma0` and `Sigma1` with the same structure: three `ROTR` operations XORed together, differing only in the rotation amounts.

$Sigma0(x) = ROTR^2(x) \oplus ROTR^{13}(x) \oplus ROTR^{22}(x)$

$Sigma1(x) = ROTR^6(x) \oplus ROTR^{11}(x) \oplus ROTR^{25}(x)$

Sigma0's pseudocode:
```pseudo
a = rotate x right by 2
b = rotate x right by 13
c = rotate x right by 22
result = a XOR b XOR c
```

As Sigma0 and Sigma1 are identical with only their constants changing, we can abstract the function to reduce duplication.

In [19]:
def sigma_rotations(x, a, b, c):
    return ROTR(x, a) ^ ROTR(x, b) ^ ROTR(x, c)


def Sigma0(x: u32) -> u32:
    """Sigma0: ROTR by 2, 13, 22, XORed together."""
    return sigma_rotations(x, 2, 13, 22)


def Sigma1(x: u32) -> u32:
    """Sigma1: ROTR by 6, 11, 25, XORed together."""
    return sigma_rotations(x, 6, 11, 25)

### sigma0(x) and sigma1(x)

[FIPS 180-4 Section 4.1.2](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf#page=15) defines `sigma0` and `sigma1` similarly to `Sigma0`/`Sigma1`, but the third operation `SHR` is shift right opposed to rotate right.

$sigma0(x) = ROTR^{7}(x) \oplus ROTR^{18}(x) \oplus SHR^{3}(x)$

$sigma1(x) = ROTR^{17}(x) \oplus ROTR^{19}(x) \oplus SHR^{10}(x)$

sigma0's pseudocode:
```pseudo
a = rotate x right by 7
b = rotate x right by 18
c = shift x right by 3
result = a XOR b XOR c
```

Same as before, we can do the same abstraction on the repeated pattern.

In [20]:
def sigma_shift(x, a, b, c):
    return ROTR(x, a) ^ ROTR(x, b) ^ SHR(x, c)

def sigma0(x: u32) -> u32:
    """sigma0: ROTR by 7, 18, then SHR by 3, XORed together."""
    return sigma_shift(x, 7, 18, 3)

def sigma1(x: u32) -> u32:
    """sigma1: ROTR by 17, 19, then SHR by 10, XORed together."""
    return sigma_shift(x, 17, 19, 10)

### Testing the Sigma and sigma functions

To assert the functions are correct we need real data for all four of the Sigma and sigma functions. Online I found this data hard to come by, except for [this SHA-256 visualiser](https://sha256algorithm.com/) ([source code](https://github.com/dmarman/sha256algorithm)). It hashes an input message step by step and shows the input and output of every function along the way.

Before trusting the tool we can check it. Hashing `abc` finishes with

`ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad`

which is the digest for `abc` given in [NIST's SHA-256 worked example](https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/SHA256.pdf), which the [Secure Hash Standard](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf) itself points to as it contains no worked examples. As the final hash is correct, we can assume the value at each step is too, and can therefore use them to test our functions.

From the visualiser, for the message `abc`:

| Function | Step on the tool | Input | Output |
|---|---|---|---|
| `sigma0` | "Calculate: w30" | `00000018` | `30060003` |
| `sigma0` | "Calculate: w31" | `61626380` | `940e90ef` |
| `sigma1` | "Calculate: w17" | `00000018` | `000f0000` |
| `sigma1` | "Calculate: w18" | `61626380` | `7da86405` |
| `Sigma0` | step 50 (round 0) | `6a09e667` | `ce20b47e` |
| `Sigma0` | step 51 (round 1) | `5d6aebcd` | `a29f64d1` |
| `Sigma1` | step 50 (round 0) | `510e527f` | `3587272b` |
| `Sigma1` | step 51 (round 1) | `fa2a4622` | `5a94fd2d` |

The test below runs each input through our function and compares the result to the tool's output. All eight rows should print `True`.

In [21]:
sigma_tests = [
    # (function, input, expected) -- values from sha256algorithm.com for "abc"
    (sigma0, 0x00000018, 0x30060003),
    (sigma0, 0x61626380, 0x940E90EF),
    (sigma1, 0x00000018, 0x000F0000),
    (sigma1, 0x61626380, 0x7DA86405),
    (Sigma0, 0x6A09E667, 0xCE20B47E),
    (Sigma0, 0x5D6AEBCD, 0xA29F64D1),
    (Sigma1, 0x510E527F, 0x3587272B),
    (Sigma1, 0xFA2A4622, 0x5A94FD2D),
]

for func, x, expected in sigma_tests:
    result = func(u32(x))
    print(f"{func.__name__}({x:08x}) = {result:08x}, "
          f"expected {expected:08x}: {result == expected}")

assert all(func(u32(x)) == expected for func, x, expected in sigma_tests)

sigma0(00000018) = 30060003, expected 30060003: True
sigma0(61626380) = 940e90ef, expected 940e90ef: True
sigma1(00000018) = 000f0000, expected 000f0000: True
sigma1(61626380) = 7da86405, expected 7da86405: True
Sigma0(6a09e667) = ce20b47e, expected ce20b47e: True
Sigma0(5d6aebcd) = a29f64d1, expected a29f64d1: True
Sigma1(510e527f) = 3587272b, expected 3587272b: True
Sigma1(fa2a4622) = 5a94fd2d, expected 5a94fd2d: True
